### ISSUE TITLE
multiple instances with different configs on one page

### ISSUE THREAD

**Issue by oudeicrat** [ofrohn/d3-celestial#96]
Is there a way of displaying multiple sky instances in different containers with different configs? I tried creating a celestial.js copy and renamed all ocurences of "Celestial" to "Celestial2nd", creating a separate celestial2ndConfig and then calling `Celestial2nd.display(celestial2ndConfig);` worked, but calling `Celestial2nd.skyview({date:myDate})` after `Celestial.skyview({date:myDate})` created a JS error saying "Celestial2nd.skyview is not a function"

**Comment by oudeicrat** [comment]
ok I discovered celestial.js is adding html elements into the main document with non-unique IDs (that might even conflict with the rest of the document) and these IDs are so interwoven throughout the code it's too much work to replace them all. I started with adding a unique idPrefix property to the Celestial2nd object defined at the top of the celestial.js file and then added this prefix to calls of .attr("id", "...") and the $ definition (`function $(id) { return document.getElementById(Celestial2nd.idPrefix+id); }`) but there are too many other ID usages, I'm not sure it's all worth it...

**Comment by oudeicrat** [comment]
well I managed to quickly patch it and so far it seems to be working, but anyway it would be nice to have an easier way of being able to use multiple instances

**Comment by ofrohn** [comment]
The configuration has a setting called 'container' for the id under which all elements of celestial are uniquely addressable if you change it to, any unique id, and name the parent div with the same id, e.g. ``#celestial-map3257 .element-class``. So b using the same Celestial with different configs it should be possible to have two maps in parallel.
But I see that I use plenty of ``d3.selectAll('.element-class')``. I can see where that would prevent multiple maps. I'll prefix all of those with the container-id. It may take some time...

**Comment by oudeicrat** [comment]
it also creates elements with specific ids, for example in the display function there is a line `if ($("loc") === null) geo(cfg);` and this "loc" element is created in the geo() function, so when one celestial instance initializes properly (creating the "loc" element), the other instance sees the "loc" element and doesn't call geo(). Among other things I added using an ID prefix to probably all usages of `.attr("id", ...)`

**Comment by nevadavid** [comment]
I found this site: https://mysky.hu/multiTervezo
I think it use this library for two or more instances. But I cannot find where is the configuration or any other related js code for working.

**Comment by nevadavid** [comment]
I tried to set multiple instances with unique id and config, but it doesn't work. The maps are loaded, but the first seems doesn't apply config.

**Comment by nevadavid** [comment]
> well I managed to quickly patch it and so far it seems to be working, but anyway it would be nice to have an easier way of being able to use multiple instances

Do you share this patch?

**Comment by oudeicrat** [comment]
> Can you share this patch?

https://github.com/oudeicrat/oudeicrat.github.io/blob/master/celestial2nd.js
used in 
https://github.com/oudeicrat/oudeicrat.github.io/blob/master/index.html

**Comment by nevadavid** [comment]
> > Can you share this patch?
> 
> https://github.com/oudeicrat/oudeicrat.github.io/blob/master/celestial2nd.js
> used in
> https://github.com/oudeicrat/oudeicrat.github.io/blob/master/index.html

Thank you @oudeicrat.
But I think it would be nice if we could load just once the celestial.js, then we use different configs for show maps.

**Comment by oudeicrat** [comment]
yes I would like that too, this was just a very quick patch to see if it is even possible at all. If I had more time I would try to make it nicer

**Comment by ofrohn** [comment]
That took some time, but I've updated the library so that every element reference is dependent on the container id (default: celestial-map, changeable in config with 'container') with the selector e.g. ``#celestial-map #elenent-id`` or ``#celestial-map .elenent-class``.
You also don't need to declare an extra div for the form, in fact, it is ignored if it is not a sibling of celestial-map, the selector-pattern is ``#celestial-map ~ #celestial-form element``.
I still need to implement multiple configurations, but with @oudeicrat's path it should work with multiple instances.
